Bottom: 8466243e5533766980a9dfaaa375adbea467c18c
Top:    73c9c51cb51a35ab76dd881d2d28ae1816708754
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2009-01-28 22:02:02 +0000

Only snap to pads on the active layer






---

diff --git a/src/crosshair.c b/src/crosshair.c
index 701bda0..58a245e 100644
--- a/src/crosshair.c
+++ b/src/crosshair.c
@@ -848,20 +848,53 @@ FitCrosshairIntoGrid (LocationType X, LocationType Y)
 
   if (PCB->RatDraw || TEST_FLAG (SNAPPINFLAG, PCB))
     ans = SearchScreenGridSlop (Crosshair.X, Crosshair.Y,
-                                PAD_TYPE | PIN_TYPE, &ptr1, &ptr2, &ptr3);
+                                PAD_TYPE, &ptr1, &ptr2, &ptr3);
   else
     ans = NO_TYPE;
 
   /* Avoid self-snapping when moving */
-  if (Settings.Mode == MOVE_MODE &&
-      Crosshair.AttachedObject.Type == ELEMENT_TYPE)
+  if (ans && Settings.Mode == MOVE_MODE &&
+      Crosshair.AttachedObject.Type == ELEMENT_TYPE &&
+      ptr1 == Crosshair.AttachedObject.Ptr1)
+    ans = NO_TYPE;
+
+  if (ans && (Settings.Mode == LINE_MODE ||
+              (Settings.Mode == MOVE_MODE &&
+               Crosshair.AttachedObject.Type == LINEPOINT_TYPE)))
     {
-      if ((ans & (PAD_TYPE | PIN_TYPE)) &&
-           ptr1 == Crosshair.AttachedObject.Ptr1)
+      PadTypePtr pad = (PadTypePtr) ptr2;
+      LayerType *desired_layer;
+      Cardinal desired_group;
+      Cardinal SLayer, CLayer;
+      int found_our_layer = False;
+
+      desired_layer = CURRENT;
+      if (Settings.Mode == MOVE_MODE &&
+          Crosshair.AttachedObject.Type == LINEPOINT_TYPE)
+        {
+          desired_layer = (LayerType *)Crosshair.AttachedObject.Ptr1;
+        }
+
+      /* find layer groups of the component side and solder side */
+      SLayer = GetLayerGroupNumberByNumber (max_layer + SOLDER_LAYER);
+      CLayer = GetLayerGroupNumberByNumber (max_layer + COMPONENT_LAYER);
+      desired_group = TEST_FLAG (ONSOLDERFLAG, pad) ? SLayer : CLayer;
+
+      GROUP_LOOP (PCB->Data, desired_group);
+      {
+        if (layer == desired_layer)
+          {
+            found_our_layer = True;
+            break;
+          }
+      }
+      END_LOOP;
+
+      if (found_our_layer == False)
         ans = NO_TYPE;
     }
 
-  if (ans & PAD_TYPE)
+  if (ans)
     {
       PadTypePtr pad = (PadTypePtr) ptr2;
       LocationType px, py;
@@ -878,7 +911,20 @@ FitCrosshairIntoGrid (LocationType X, LocationType Y)
           nearest = sq_dist;
         }
     }
-  else if (ans & PIN_TYPE)
+
+  if (PCB->RatDraw || TEST_FLAG (SNAPPINFLAG, PCB))
+    ans = SearchScreenGridSlop (Crosshair.X, Crosshair.Y,
+                                PIN_TYPE, &ptr1, &ptr2, &ptr3);
+  else
+    ans = NO_TYPE;
+
+  /* Avoid self-snapping when moving */
+  if (ans && Settings.Mode == MOVE_MODE &&
+      Crosshair.AttachedObject.Type == ELEMENT_TYPE &&
+      ptr1 == Crosshair.AttachedObject.Ptr1)
+    ans = NO_TYPE;
+
+  if (ans)
     {
       PinTypePtr pin = (PinTypePtr) ptr2;
       sq_dist = SQUARE (pin->X - Crosshair.X) + SQUARE (pin->Y - Crosshair.Y);
